#4 - Implementar rota que permita a atualização das informações de um usuário

atenokida opened the pull request:
## Descrição
Implementado endpoint para atualizar informações de usuário.


## Contexto
motivação: Implementar atualização de informação de usuário.

## Como testar

Enviar requisição ``PATCH`` para o path: ``/users/{id}/update`` onde ``{id}`` é o uuid do usuário.
<hr>

Campos (opcionais): ``name``; ``email``; ``phone``.

Exemplo de ``body``:
```
{
    "name": "John doe",
    "email": "john.doe@email.com",
    "phone": "+5555112345678",
}
```

<hr>


### Escritor
- [x] A rota `/users/{id}/update` está funcionando
  - [x] As validações estão funcionando
    - [x] Caso o uuid seja inválido lança exceção `400`
    - [x] As validações de nome estão funcionando
      - [x] Caso o nome seja inválido lança exceção `400`
    - [x] As validações de email estão funcionando
      - [x] Caso o email seja inválido lança exceção `400`
      - [x] Caso o email já exista lança exceção `500`
    - [x] As validações de telefone estão funcionando
      - [x] Caso o telefone seja inválido lança exceção `400`
      - [x] Caso o telefone já exista lança exceção `500`




### Tester
- [x] A rota `/users/{id}/update` está funcionando
  - [x] As validações estão funcionando
    - [x] Caso o uuid seja inválido lança exceção `400`
    - [x] As validações de nome estão funcionando
      - [x] Caso o nome seja inválido lança exceção `400`
    - [x] As validações de email estão funcionando
      - [x] Caso o email seja inválido lança exceção `400`
      - [x] Caso o email já exista lança exceção `500`
    - [x] As validações de telefone estão funcionando
      - [x] Caso o telefone seja inválido lança exceção `400`
      - [x] Caso o telefone já exista lança exceção `500`

## Natureza da alteração
<!-- Indica o tipo de alteração que foi feita -->
- [x] Nova Feature.
- [ ] Correção de BUG.
- [ ] Refatoração de Código.

atenokida committed:
feat: Initial commit for endpoint to update a new user

atenokida committed:
Merge conflicts with master

atenokida committed:
feat: Implement UpdateClient method on users-service

atenokida committed:
feat: Implement UpdateUserById on users-controller

atenokida committed:
fix: Remove unused fields on UpdateUserInput in my-types

atenokida committed:
docs: Update CHANGELOG.md

afmireski commented on internal/ports/user-repository-port.go:
```diff
@@ -18,10 +18,17 @@ type UserRepositoryPort interface {
+	Password      string    `json:"password"`
+	PlainPassword string    `json:"plain_password"`
+	BirthDate     time.Time `json:"birth_date"`
+}
```
Apesar do repository deixar, não faz sentido atualizar a senha por um update geral no service. Geralmente é criada uma função só para isso, pois além de fornecer a nova senha, o usuário tem que mandar a confirmação.
Inclusive esses campos nem estão sendo utilizados no seu service.

afmireski commented on internal/ports/user-repository-port.go:
Olhando a indicação de alteração, aparentemente vai ter uma perca de código já pronto nesse arquivo, no que diz respeito ao CreateUserInput.
Provavelmente algum detalhe foi modificado na struct e o git acha que é para alterá-la. Sempre tomar cuidado quando alterar código que vêm da main.

afmireski commented on internal/services/users-service.go:
```diff
@@ -77,6 +76,38 @@ func (s *UsersService) NewUser(input myTypes.NewUserInput) *customErrors.Interna
 	return nil
 }
 
+func (s *UsersService) UpdateClient(id string, input myTypes.UpdateUserInput) *customErrors.InternalError {
```
Varia de caso a caso, mas num `update` até que é interessante retornar o objeto atualizado, elimina a necessidade do Front ter que fazer uma nova busca assim que a operação é realizada.

afmireski commented on internal/web/controllers/users-controller.go:
```diff
@@ -45,6 +45,33 @@ func (c *UsersController) NewUser(w http.ResponseWriter, r *http.Request) {
+		return
+	}
+
+	w.WriteHeader(http.StatusCreated)
```
Como nenhum dado novo é criado, você está atualizando um usuário, o status `201` não é adequado. Ele geralmente está relacionado a Posts que criam uma informação nova.  
Em outros métodos, como delete ou update, caso nada seja retornado é melhor usado o status `204, No Content` .

afmireski commented on internal/web/controllers/users-controller.go:
```diff
@@ -45,6 +45,33 @@ func (c *UsersController) NewUser(w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(http.StatusCreated)
 }
 
+func (c *UsersController) UpdateUserById(w http.ResponseWriter, r *http.Request) {
```
Nesse caso não precisa especificar que é por id o update, no momento só temos esse método que faz isso para os usuários. Como vamos ter os adms depois, talvez seja interessante renomear para `UpdateClient`, mas é só uma sugestão. `UpdateUser` já deixa bem claro o propósito.
```suggestion
func (c *UsersController) UpdateUser(w http.ResponseWriter, r *http.Request) {
```

afmireski commented on CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@
 * [#16](https://github.com/afmireski/garchop-api/issues/16) - Implementar rota de Login na plataforma
     * Refatoração do cadastro de usuário para registrar usuário na tabela de autenticação do supabase.
 * [#5](https://github.com/afmireski/garchop-api/issues/5) - Implementar rota para excluir um usuário
+* [#4](https://github.com/afmireski/garchop-api/issues/4) - Implementar rota que permita a atualização das informações de um usuário
 * [#3](https://github.com/afmireski/garchop-api/issues/3) - Implementar rota que possibilite o cadastro de um usuário
 * [#2](https://github.com/afmireski/garchop-api/issues/2) - Implementar rota para buscar informações de um usuário
```
O Changelog geralmente segue a ordem de implementação das tarefas, não necessariamente uma sequência numérica. Então como essa Issue, 4, foi implementada depois da 16, o certo seria ela ficar no topo.
```suggestion
* [#4](https://github.com/afmireski/garchop-api/issues/4) - Implementar rota que permita a atualização das informações de um usuário
* [#16](https://github.com/afmireski/garchop-api/issues/16) - Implementar rota de Login na plataforma
    * Refatoração do cadastro de usuário para registrar usuário na tabela de autenticação do supabase.
* [#5](https://github.com/afmireski/garchop-api/issues/5) - Implementar rota para excluir um usuário
* [#3](https://github.com/afmireski/garchop-api/issues/3) - Implementar rota que possibilite o cadastro de um usuário
* [#2](https://github.com/afmireski/garchop-api/issues/2) - Implementar rota para buscar informações de um usuário
```

afmireski reviewed (changes requested):
Apontei algumas sugestões. Só dar uma revisada nas structs e está tudo certo.

atenokida committed:
docs: change task #4 sort

atenokida committed:
refactor: Change `UpdateUserById` to `UpdateClient`

atenokida committed:
fix: Change http return type for update from 201 to 204

atenokida committed:
fix(supabase-users-repository): Date type treatment on SerializeMany()

atenokida committed:
feat: Add `Password` field on users model

atenokida committed:
feat: Return updated obj on controller

atenokida committed:
feat: Return obj on users service UpdateClient()

Returns input object case fails to update

atenokida committed:
feat: Add mapToUpdatedUserInput() method + imports

atenokida committed:
fix: Only update if field exists

Previously all three fields (name, email, phone) we're being sent to the repository update method even if they didn't exist on the json.

atenokida committed:
fix: Remove unused UsersUpdateInput dto

Duplicate of /internal/types/input-types.go/NewUserInput

afmireski reviewed (approved):
Tudo certo com o código.

afmireski committed:
#4 - Fix UpdateClient response serialization
